kaldb/src/main/java/com/slack/kaldb/clusterManager/ClusterHpaMetricService.java

```diff
@@ -0,0 +1,172 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The Cluster HPA metrics service is intended to be run on the manager node, and is used for making
+ * centralized, application-aware decisions that can be used to inform a Kubernetes horizontal pod
+ * autoscaler (HPA).
+ *
+ * @see <a
+ *     href="https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/">Kubernetes
+ *     HPA</a>
+ */
+public class ClusterHpaMetricService extends AbstractScheduledService {
+  private static final Logger LOG = LoggerFactory.getLogger(ClusterHpaMetricService.class);
+
+  // todo - consider making over-provision and lock duration configurable values
+  protected int CACHE_SLOT_OVER_PROVISION = 1000;
+  protected Duration CACHE_SCALEDOWN_LOCK = Duration.of(15, ChronoUnit.MINUTES);
+
+  private final ReplicaMetadataStore replicaMetadataStore;
+  private final CacheSlotMetadataStore cacheSlotMetadataStore;
+  private final HpaMetricMetadataStore hpaMetricMetadataStore;
+  protected final Map<String, Instant> cacheScalingLock = new ConcurrentHashMap<>();
+  protected static final String CACHE_HPA_METRIC_NAME = "hpa_cache_demand_factor_%s";
+
+  public ClusterHpaMetricService(
+      ReplicaMetadataStore replicaMetadataStore,
+      CacheSlotMetadataStore cacheSlotMetadataStore,
+      HpaMetricMetadataStore hpaMetricMetadataStore) {
+    this.replicaMetadataStore = replicaMetadataStore;
+    this.cacheSlotMetadataStore = cacheSlotMetadataStore;
+    this.hpaMetricMetadataStore = hpaMetricMetadataStore;
+  }
+
+  @Override
+  protected Scheduler scheduler() {
+    return Scheduler.newFixedRateSchedule(Duration.ofSeconds(15), Duration.ofSeconds(30));
+  }
+
+  @Override
+  protected void runOneIteration() {
+    LOG.info("Running ClusterHpaMetricService");
+    try {
+      publishCacheHpaMetrics();
+    } catch (Exception e) {
+      LOG.error("Error running ClusterHpaMetricService", e);
+    }
+  }
+
+  /**
+   * Calculates and publishes HPA scaling metric(s) to Zookeeper for the Cache nodes. This makes use
+   * of the Kubernetes HPA formula, which is:
+   *
+   * <p>desiredReplicas = ceil[currentReplicas * ( currentMetricValue / desiredMetricValue )]
+   *
+   * <p>Using this formula, if we inform the user what value to set for the desiredMetricValue, we
+   * can vary the currentMetricValue to control the scale of the cluster. This is accomplished by
+   * producing a metric (hpa_cache_demand_factor_REPLICASET) that targets a value of 1.0. As long as
+   * the HPA is configured to target a value of 1.0 this will result in an optimized cluster size
+   * based on cache slots vs replicas, plus a small buffer.
+   *
+   * <pre>
+   * metrics:
+   *   - type: Pods
+   *     pods:
+   *       metric:
+   *         name: hpa_cache_demand_factor_REPLICASET
+   *       target:
+   *         type: AverageValue
+   *         averageValue: 1.0
+   * </pre>
+   */
+  private void publishCacheHpaMetrics() {
+    Set<String> replicaSets =
```

Comment:
qq: do we need `listSync` in all these places? I presume yes, but maybe we can add a small statement as to why?

Reply:
After the ZK refactor, `listSync` uses the cache now. The only other option here is the listAsync but that's more of a syntax preference if you want to use the completable future or not.